[MSE][GStreamer] Workaround basesink's lack of support for position queries between a non-resetting flush and a pre-roll

ntrrgc opened the pull request:
#### 080880118a0748f903a6cdc0f69546f868533bf2
<pre>
[MSE][GStreamer] Workaround basesink&apos;s lack of support for position queries between a non-resetting flush and a pre-roll
<a href="https://bugs.webkit.org/show_bug.cgi?id=248683">https://bugs.webkit.org/show_bug.cgi?id=248683</a>

Reviewed by Xabier Rodriguez-Calvar.

Workaround for upstream GStreamer patch projected to land in GStreamer
1.24.0: <a href="https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/3471">https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/3471</a>
basesink: Support position queries after non-resetting flushes

Briefly explained, before the fix GStreamer sinks will fail to answer
position queries after a FLUSH_STOP(reset_time=FALSE) until they receive
a new segment, even though the sinks in this situation will preserve the
previous segment.

This is particularly a problem for streams with quality changes, such as
MSE, but potentially also native adaptive streaming implementations, as
the currentTime will suddenly reset to zero, and unlike seeks, quality
change flushes generally occur automatically without user intervention.

The bug means that currentTime can reset to zero seemingly randomly when
a stream is flushed for reason other than a seek, notably in MSE, and
most notably after <a href="https://github.com/WebKit/WebKit/pull/3966">https://github.com/WebKit/WebKit/pull/3966</a> finally
gets landed. This has been seen in failures for the test
media/media-source/media-source-seek-unbuffered.html -- although that
test has several simultaneous unrelated issues that need work on.

To demonstrate the issue in a more isolated case, a new test case has
been added:

imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime.html

As of writing there are no plans currently for the patch to be
officially backported into the GStreamer 1.22.x series due to concerns
over making changes on critical paths of critical elements.

Therefore, in order to not suffer this problem in WebKit with any
currently released versions of GStreamer, we need a workaround.

This is a difficult bug to work around: you can add probes that capture
queries flowing through a pipeline, but you cannot add a probe that
captures a query that is sent directly to an element without going
through any pads (e.g. with gst_element_query_position()). Unfortunately
this is more often than not how position queries are made.

Since we cannot monkey-patch or probe the position queries for basesink
directly, the workaround is instead based on analysis and manipulation
of the basesink internals so that gst_base_sink_get_position() won&apos;t
fail in the desired case, by tweaking basesink-&gt;have_newsegment inside a
probe for the FLUSH_STOP event.

I went through lots of different approaches for this workaround over
weeks, and the workaround method proposed was the final one that was
able to fix the issue without introducing any regressions in
LayoutTests.

This is the second workaround for an issue in GStreamer sinks that takes
advantage of re-registering elements. To keep things clean, both
workarounds have been grouped in a new file
GStreamerSinksWorkarounds.cpp, and refactored into classes with similar
interfaces.

Also, now the workarounds can be enabled and disabled independently with
new environment variables:

* WEBKIT_GST_WORKAROUND_APP_SINK_FLUSH_CAPS
* WEBKIT_GST_WORKAROUND_BASE_SINK_POSITION_FLUSH

Allowed values are: ForceEnable, ForceDisabled and UseIfNeeded
(case-insensitive). The latter enables each workaround only for
GStreamer versions known to need it.

UseIfNeeded is the default value for both workarounds, but this can be
changed at build time if desired by setting the following macros:

* WEBKIT_GST_WORKAROUND_APP_SINK_FLUSH_CAPS_DEFAULT_MODE
* WEBKIT_GST_WORKAROUND_BASE_SINK_POSITION_FLUSH_DEFAULT_MODE

* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime-expected.txt: Added.
* Source/WebCore/platform/GStreamer.cmake:
* Source/WebCore/platform/graphics/gstreamer/AppSinkWorkaround.cpp: Removed.
* Source/WebCore/platform/graphics/gstreamer/AppSinkWorkaround.h: Removed.
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr&lt;GstBaseSink&gt;):
(WTF::derefGPtr&lt;GstBaseSink&gt;):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::ensureGStreamerInitialized):
* Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp: Added.
(WebCore::getWorkAroundModeFromEnvironment):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::isNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::installIfNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::checkIsNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::initializeIsNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::flushIsNonResetting):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::probe):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::newUserData):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::deleteUserData):
(WebCore::AppSinkFlushCapsWorkaroundProbe::isNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::installIfNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::checkIsNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::initializeIsNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::probe):
(WebCore::AppSinkFlushCapsWorkaroundProbe::newUserData):
(WebCore::AppSinkFlushCapsWorkaroundProbe::deleteUserData):
(WebCore::registerAppsinkWithWorkaroundsIfNeededCallOnce):
(WebCore::registerAppsinkWithWorkaroundsIfNeeded):
(WebCore::installBaseSinkPositionFlushWorkaroundIfNeeded):
(webkitAppSinkWithWorkAroundsConstructed):
(webkit_app_sink_with_workarounds_class_init):
* Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.h: Added.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::gstreamerPositionFromSinks const):
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcStreamFlush):
* Tools/Scripts/webkitpy/style/checker.py:

Canonical link: <a href="https://commits.webkit.org/264476@main">https://commits.webkit.org/264476@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f37172ef5f6f62a3b9c0ddd6cd20e2b89a4f017c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7743 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9384 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7897 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7748 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10004 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7935 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10769 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7877 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7108 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9493 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/7859 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6300 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7047 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7451 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7173 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10580 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7666 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6249 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6993 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6957 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1848 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11205 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7398 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime.html:
```diff
@@ -0,0 +1,69 @@
+function waitForCurrentTimeToReach(test, mediaElement, targetTime, callback)
+{
+    let lastTime = mediaElement.currentTime;
+    let timeUpdateHandler = test.step_func(function _timeUpdateHandler() {
```
This can probably be a const, can't it?

ntrrgc replied:
`timeUpdateHandler` can indeed be const.

calvaris commented on LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime.html:
```diff
@@ -0,0 +1,69 @@
+    mediaElement.addEventListener('timeupdate', timeUpdateHandler);
+}
+
+function testCurrentTimeDoesntRegress(test, mediaElement, initialTime, callback)
```
Since we can't use apostrophes, better to write DoesNot

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+private:
+    static bool s_isNeeded;
+
+    bool m_handlingFlushStop { false };
```
Lacks an are or is in the name to comply with the naming style of booleans.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+
+    static bool checkIsNeeded()
+    {
+        GST_DEBUG("BaseSinkPositionFlushWorkaroundProbe: running GStreamer %s, the bug fix is expected to land in 1.24.0.", gst_version_string());
```
I guess we'll have to follow up 1.24.0 to ensure it is there, otherwise we'll have to change this.

ntrrgc replied:
It should be there unless somehow reverted.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+    static GstPadProbeReturn probe(GstPad* pad, GstPadProbeInfo* info, void* userData)
+    {
+        auto* self = static_cast<BaseSinkPositionFlushWorkaroundProbe*>(userData);
+        GstBaseSink* basesink = GST_BASE_SINK(GST_PAD_PARENT(pad));
```
I think we should use `gst_pad_get_parent` and adopt the result. IMHO it's a better practice.

ntrrgc replied:
Using `gst_pad_get_parent()` would unnecessarily lock the `GObject` mutex, as the pad is not going to be detached from the object... But I have checked that `probe_hook_marshal()` unlocks it, so it should be fine to use `gst_pad_get_parent()` as well.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+        auto* self = static_cast<BaseSinkPositionFlushWorkaroundProbe*>(userData);
+        GstBaseSink* basesink = GST_BASE_SINK(GST_PAD_PARENT(pad));
+        GST_TRACE_OBJECT(pad, "m_handlingFlushStop: %s, pad is flushing: %s, received: %" GST_PTR_FORMAT, boolForPrinting(self->m_handlingFlushStop), boolForPrinting(GST_PAD_IS_FLUSHING(pad)), info->data);
+        if (!self->m_handlingFlushStop && (info->type & GST_PAD_PROBE_TYPE_EVENT_FLUSH) && GST_EVENT_TYPE(info->data) == GST_EVENT_FLUSH_STOP
```
Are the () needed around the binary & ?

ntrrgc replied:
It doesn't matter what the C++ language spec says. If you ever find yourself asking whether parentheses are necessary, the answer is always yes.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+            GST_DEBUG_OBJECT(pad, "Non-resetting FLUSH_STOP has been handled by this element and its downstream. Now force position queries to work by setting basesink->have_newsegment to TRUE.");
+            basesink->have_newsegment = TRUE;
+
+            return GST_PAD_PROBE_HANDLED;
```
We seem to be leaking the event. According to the doc, when you return this, you need to free the event. Maybe you want to switch to `DROP`.

ntrrgc replied:
Where did you see this? gst_pad_send_event() takes `[transfer full]` so we shouldn't be leaking...

calvaris replied:
You're right, there is no leak because of that transfer full that I missed. But not that you mention it, it would be better to use `GST_PAD_PROBE_INFO_EVENT` instead of `GST_EVENT(info->data)`

ntrrgc replied:
Oh. TIL about GST_PAD_PROBE_INFO_EVENT. Cool.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+        // Changes to the flushing flag of a pad only occur while the pad lock is held.
+        // By holding it, we can reliably prevent the flushing flag from changing during the execution of our code.
+        // The pad lock also ensures there are no data races over `priv->needsResendCaps`.
+        GST_OBJECT_LOCK(pad);
```
You can use `GstObjectLocker`

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+            GST_DEBUG_OBJECT(pad, "Buffer received, but first need to resend pad caps to workaround bug. Will resend caps.");
+            willResendCaps = true;
+        }
+        GST_OBJECT_UNLOCK(pad);
```
and `unlockEarly` here.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+            GRefPtr<GstCaps> caps = adoptGRef(gst_pad_get_current_caps(pad));
+            GST_DEBUG_OBJECT(pad, "Sending stored pad caps to appsink: %" GST_PTR_FORMAT, caps.get());
+            // This will cause a recursive call to appsinkWorkaroundProbe() which will also set `needsResendCaps` to false.
+            bool capsSent = gst_pad_send_event(pad, gst_event_new_caps(caps.get()));
```
`areCapsSent`

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+
+static void registerAppsinkWithWorkaroundsIfNeededCallOnce()
+{
+    // If any workarounds are needed in this system, override GStreamer appsink for a version containing any needed workarounds
```
.

This is bait, right?

ntrrgc replied:
No, I have bigger worries in my mind than the periods at the end of single-sentence comments.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,331 @@
+    // If any workarounds are needed in this system, override GStreamer appsink for a version containing any needed workarounds
+    GST_DEBUG_CATEGORY_INIT(webkit_workarounds_debug, "webkitworkarounds", 0, "WebKit GStreamer Workarounds");
+    GST_DEBUG("Checking for potentially needed GStreamer workarounds...");
+    bool needsWorkarounds = BaseSinkPositionFlushWorkaroundProbe::isNeeded() || AppSinkFlushCapsWorkaroundProbe::isNeeded();
```
`doesNeedWorkarounds`

calvaris reviewed (approved):
Nice work! Only some cosmetic changes needed.

ntrrgc committed:
[MSE][GStreamer] Workaround basesink's lack of support for position queries between a non-resetting flush and a pre-roll
https://bugs.webkit.org/show_bug.cgi?id=248683

Reviewed by Xabier Rodriguez-Calvar.

Workaround for upstream GStreamer patch projected to land in GStreamer
1.24.0: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/3471
basesink: Support position queries after non-resetting flushes

Briefly explained, before the fix GStreamer sinks will fail to answer
position queries after a FLUSH_STOP(reset_time=FALSE) until they receive
a new segment, even though the sinks in this situation will preserve the
previous segment.

This is particularly a problem for streams with quality changes, such as
MSE, but potentially also native adaptive streaming implementations, as
the currentTime will suddenly reset to zero, and unlike seeks, quality
change flushes generally occur automatically without user intervention.

The bug means that currentTime can reset to zero seemingly randomly when
a stream is flushed for reason other than a seek, notably in MSE, and
most notably after https://github.com/WebKit/WebKit/pull/3966 finally
gets landed. This has been seen in failures for the test
media/media-source/media-source-seek-unbuffered.html -- although that
test has several simultaneous unrelated issues that need work on.

To demonstrate the issue in a more isolated case, a new test case has
been added:

imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime.html

As of writing there are no plans currently for the patch to be
officially backported into the GStreamer 1.22.x series due to concerns
over making changes on critical paths of critical elements.

Therefore, in order to not suffer this problem in WebKit with any
currently released versions of GStreamer, we need a workaround.

This is a difficult bug to work around: you can add probes that capture
queries flowing through a pipeline, but you cannot add a probe that
captures a query that is sent directly to an element without going
through any pads (e.g. with gst_element_query_position()). Unfortunately
this is more often than not how position queries are made.

Since we cannot monkey-patch or probe the position queries for basesink
directly, the workaround is instead based on analysis and manipulation
of the basesink internals so that gst_base_sink_get_position() won't
fail in the desired case, by tweaking basesink->have_newsegment inside a
probe for the FLUSH_STOP event.

I went through lots of different approaches for this workaround over
weeks, and the workaround method proposed was the final one that was
able to fix the issue without introducing any regressions in
LayoutTests.

This is the second workaround for an issue in GStreamer sinks that takes
advantage of re-registering elements. To keep things clean, both
workarounds have been grouped in a new file
GStreamerSinksWorkarounds.cpp, and refactored into classes with similar
interfaces.

Also, now the workarounds can be enabled and disabled independently with
new environment variables:

* WEBKIT_GST_WORKAROUND_APP_SINK_FLUSH_CAPS
* WEBKIT_GST_WORKAROUND_BASE_SINK_POSITION_FLUSH

Allowed values are: ForceEnable, ForceDisabled and UseIfNeeded
(case-insensitive). The latter enables each workaround only for
GStreamer versions known to need it.

UseIfNeeded is the default value for both workarounds, but this can be
changed at build time if desired by setting the following macros:

* WEBKIT_GST_WORKAROUND_APP_SINK_FLUSH_CAPS_DEFAULT_MODE
* WEBKIT_GST_WORKAROUND_BASE_SINK_POSITION_FLUSH_DEFAULT_MODE

* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/media-source/mediasource-remove-preserves-currentTime-expected.txt: Added.
* Source/WebCore/platform/GStreamer.cmake:
* Source/WebCore/platform/graphics/gstreamer/AppSinkWorkaround.cpp: Removed.
* Source/WebCore/platform/graphics/gstreamer/AppSinkWorkaround.h: Removed.
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr<GstBaseSink>):
(WTF::derefGPtr<GstBaseSink>):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::ensureGStreamerInitialized):
* Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp: Added.
(WebCore::getWorkAroundModeFromEnvironment):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::isNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::installIfNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::checkIsNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::initializeIsNeeded):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::flushIsNonResetting):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::probe):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::newUserData):
(WebCore::BaseSinkPositionFlushWorkaroundProbe::deleteUserData):
(WebCore::AppSinkFlushCapsWorkaroundProbe::isNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::installIfNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::checkIsNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::initializeIsNeeded):
(WebCore::AppSinkFlushCapsWorkaroundProbe::probe):
(WebCore::AppSinkFlushCapsWorkaroundProbe::newUserData):
(WebCore::AppSinkFlushCapsWorkaroundProbe::deleteUserData):
(WebCore::registerAppsinkWithWorkaroundsIfNeededCallOnce):
(WebCore::registerAppsinkWithWorkaroundsIfNeeded):
(WebCore::installBaseSinkPositionFlushWorkaroundIfNeeded):
(webkitAppSinkWithWorkAroundsConstructed):
(webkit_app_sink_with_workarounds_class_init):
* Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.h: Added.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::gstreamerPositionFromSinks const):
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(webKitMediaSrcStreamFlush):
* Tools/Scripts/webkitpy/style/checker.py:

Canonical link: https://commits.webkit.org/264476@main

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
```diff
@@ -0,0 +1,333 @@
+        if (!isNeeded())
+            return;
+
+        GRefPtr<GstPad> pad = gst_element_get_static_pad(GST_ELEMENT(appsink), "sink");
```
missing `adoptGRef()` here

calvaris replied:
TRUE!

ntrrgc replied:
Good catch. I can make a follow-up patch to fix it.
